### ISSUE TITLE
eDonkey dissector

### ISSUE THREAD

**Issue by 41CEA55** [ntop/nDPI#2408]
**Is your feature request related to a problem? Please describe.**

Yea, eDonkey protocol dissector looks broken and doesn't classify UDP traffic.

[edonkey_udp.zip](https://github.com/ntop/nDPI/files/15084224/edonkey_udp.zip)

**Describe the solution you'd like**

I'd suggest removing dissector, but eDonkey still works even to this day - yea, the [number of servers](https://www.emule-security.org/serverlist/) can be counted on your fingers, but you can still download some old movies, software and porn there.

So I suggest to keep only eDonkey over TCP classification, since eDonkey over UDP classification would be quite problematic to implement (shit, the minimum payload length of a valid packet is 2 bytes!). In the case of eDonkey over TCP, it would be enough to perform such a simple check:

```
if (packet->tcp != NULL && packet->payload_packet_len > 5)
{
  /* 0xE3 - eDonkey, 0xC5 - eMule extensions, 0xD4 - eMule compressed */
  u_int8_t protocol = packet->payload[0];
  if (protocol != 0xE3 && protocol != 0xC5 && protocol != 0xD4) {
    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
    return;
  }
  /* All eDonkey over TCP packets should have this field */
  u_int32_t message_length = packet->payload_packet_len-5;
  if (message_length == le32toh(get_u_int32_t(packet->payload, 1))) {
    ndpi_int_edonkey_add_connection(ndpi_struct, flow);
    return;
  }
}

NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
```

I don't think this will cause false positives since the first byte has only 3 valid values and the next 4 bytes store the TCP payload length minus the eDonkey header length in little endian.

@IvanNardi @utoni what do you think, guys?

**Comment by IvanNardi** [comment]
I agree. we should remove UDP and update TCP: your code seems a good starting point
